### PR TITLE
samples: nrf9160: modem_shell: add MQTT connection to nRF cloud

### PR DESCRIPTION
**Pull request by hiltunent**
New commands "cloud connect" and "cloud disconnect" added. Enable
by setting CONFIG_MOSH_CLOUD. Establishes MQTT connection to nRF
Cloud.

Signed-off-by: Tuomas Hiltunen <tuomas.hiltunen@nordicsemi.no>

### CODE REVIEW

File: samples/nrf9160/modem_shell/prj.conf [hunk 1]
```diff
@@ -128,6 +133,8 @@ CONFIG_MULTICELL_LOCATION_MAX_NEIGHBORS=17
 
 # Cellular positioning services
 CONFIG_MULTICELL_LOCATION_SERVICE_NRF_CLOUD=y
+# Uncomment the following if CONFIG_NRF_CLOUD_MQTT is enabled
```
Comment (jhirsi) [inline samples/nrf9160/modem_shell/prj.conf:136]: Somehow, cannot do that just in this PR, we should eventually make these configurations easier between MQTT and REST.

At here, maybe worth to note that basically it is either MQTT or REST for the nrf cloud services. 

Also was thinking if CONFIG_NRF_CLOUD_CELL_POS should have CONFIG_NRF_CLOUD_MQTT as a dependency, but that is not a subject of this PR.

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef CLOUD_SHELL_H
+#define CLOUD_SHELL_H
+
+int cloud_shell(const struct shell *shell, size_t argc, char **argv);
+
+#endif /* SREST_CLOUD_SHELL_H */
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.h:12]:
```suggestion
#endif /* CLOUD_SHELL_H */
```

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+	"  connect:      Establish MQTT connection to nRFCloud.\n"
+	"  disconnect:   Disconnect from nRFCloud.\n";
+
+static void cloud_shell_print_usage(const struct shell *shell)
+{
+	mosh_print_no_format(cloud_shell_cmd_usage_str);
+}
+
+#if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+static int send_service_info(void)
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:42]: I'm not sure if it's really necessary to update the device shadow from modem_shell. And why it's tied to A-GPS/P-GPS configuration?

My guess is that it isn't necessary to update the shadow, because modem_shell doesn't send the location to the cloud anyway. It looks like this tells the cloud that the cloud UI should display the GPS location window. So, because we don't send the location to the cloud, this shouldn't be necessary. And if we still want to do that, then it probably shouldn't depend on the A-GPS/P-GPS flags but should be done always.

File: samples/nrf9160/modem_shell/prj.conf [hunk 4]
```diff
@@ -118,6 +119,10 @@ CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS=0
 CONFIG_NRF_CLOUD_AGPS=y
 CONFIG_NRF_CLOUD_REST=y
 CONFIG_MODEM_JWT=y
+# Uncomment the following if CONFIG_MOSH_CLOUD is enabled
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/prj.conf:122]: Makes me wonder if there should be yet another overlay for MQTT. :)

File: samples/nrf9160/modem_shell/Kconfig [hunk 5]
```diff
@@ -84,6 +84,13 @@ config MOSH_REST
 	help
 	  REST client for sending and receiving simple REST requests/responses.
 
+config MOSH_CLOUD
+	bool "MQTT connection to nRFCloud"
+	default y
+	select CLOUD
```
[the remark this answers is not in the document]
Comment (hiltunent) [reply samples/nrf9160/modem_shell/Kconfig:90]: Hmm, dunno where that "select CLOUD" came from. I'll remove it.
I guess I not consider all the implications setting this to 'y' has on other commands, good point.
I agree the name should be more specific...should we even mention NRF in the name? Is it likely that there will be other optional clouds to use here?

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c (hunk 3 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:42]: You are right. This doesn't seem to do anything at the moment. I'll remove it.

File: samples/nrf9160/modem_shell/prj.conf (hunk 4 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/prj.conf:122]: Actually I found out that you don't even need to set CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD=y and CONFIG_MQTT_KEEPALIVE=1200 as they are already default values so I think we don't need an another overlay quite yet :)

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	err = nrf_cloud_connect(NULL);
+	if (err == NRF_CLOUD_CONNECT_RES_ERR_ALREADY_CONNECTED) {
+		mosh_print("nRF Cloud connection already established");
+	} else if (err) {
+		mosh_error("nrf_cloud_connect, error: %d", err);
+	} else {
+		mosh_print("********************************************");
+		mosh_print(" Connection to nRF Cloud established");
+		mosh_print(" Endpoint:    %s", CONFIG_NRF_CLOUD_HOST_NAME);
+		mosh_print("********************************************");
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:201]: Are we too visible here. While I don't have anything against this in general, we haven't used such a highlighted style anywhere else.

File: samples/nrf9160/modem_shell/src/shell.c [hunk 7]
```diff
@@ -223,3 +227,9 @@ SHELL_CMD_REGISTER(rest, NULL,
 	"REST client.",
 	rest_shell);
 #endif
+
+#if defined(CONFIG_MOSH_CLOUD)
+SHELL_CMD_REGISTER(cloud, NULL,
+	"MQTT connection to nRFCloud.",
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/shell.c:233]: We should have a discussion on the scope of "cloud" command. Whether it's intended just for nRF Cloud or whether it could be extended later. No changes required here I guess if we don't plan to make the command "nrf_cloud".

Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/shell.c:233]: agree. I have had thoughts for this command in a widers scope, i.e. beyond mqtt and nrf cloud. ...but maybe later.

File: samples/nrf9160/modem_shell/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/Kconfig]: I think the name can be as it is for now.

File: samples/nrf9160/modem_shell/prj.conf (hunk 1 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/prj.conf:136]: I totally agree, MQTT / REST configuration is a mess now.

I'll improve the comment.

Yes, CONFIG_NRF_CLOUD_CELL_POS should have CONFIG_NRF_CLOUD_MQTT as a dependency. It won't even compile if  CONFIG_NRF_CLOUD_CELL_POS  is enabled and CONFIG_NRF_CLOUD_MQTT is not.

File: samples/nrf9160/modem_shell/prj.conf [hunk 8]
```diff
@@ -118,6 +119,10 @@ CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS=0
...
+#CONFIG_NRF_CLOUD_MQTT=y
+#CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD=y
+#CONFIG_MQTT_KEEPALIVE=1200
```
[the remark this answers is not in the document]
Comment (hiltunent) [reply samples/nrf9160/modem_shell/prj.conf:125]: ok, also NRF_CLOUD_CONNECTION_POLL_THREAD=y is default

File: samples/nrf9160/modem_shell/src/shell.c (hunk 7 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/src/shell.c:233]: Some thoughts: if we plan to add other clouds, I would make the command something like "cloud connect amazon". In this case, the current naming would be ok I think. And if we don't add other clouds, then there is no point to change the name to nrf_cloud as there is no way to confuse it with any other cloud :)

File: samples/nrf9160/modem_shell/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (hiltunent) [reply samples/nrf9160/modem_shell/Kconfig]: set to 'n' by default

File: samples/nrf9160/modem_shell/Kconfig [hunk 9]
```diff
@@ -84,6 +84,13 @@ config MOSH_REST
 	help
 	  REST client for sending and receiving simple REST requests/responses.
 
+config MOSH_CLOUD
+	bool "MQTT connection to nRFCloud"
+	default n
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/Kconfig:89]: There's no need for "default n", because that's the default value. Somebody will anyway ask you to remove this line, so it may as well be me. :)

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+		}
+#else
+		if (err) {
+			mosh_print("Unable to process A-GPS data, error: %d", err);
+		}
+#endif
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST:
+		mosh_print("NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST");
+		mosh_warn("Add the device to nRF Cloud and wait for it to reconnect");
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:111]: Does the nRF Cloud library trigger reconnect automatically or is this something that the application needs to do?

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+#endif
+
+	err = nrf_cloud_connect(NULL);
+	if (err == NRF_CLOUD_CONNECT_RES_ERR_ALREADY_CONNECTED) {
+		mosh_print("nRF Cloud connection already established");
+	} else if (err) {
+		mosh_error("nrf_cloud_connect, error: %d", err);
+	} else {
+		mosh_print("Connection to nRF Cloud established");
+		mosh_print("Endpoint: %s", CONFIG_NRF_CLOUD_HOST_NAME);
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:151]: I'd make these into one line print unless there is specific reason to have two lines?

File: samples/nrf9160/modem_shell/src/shell.c (hunk 7 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/shell.c:233]: Yep, as a command name, `cloud` is most future-proof. And I think the current usage is fine. However, I reserve right to suggest backwards incompatible changes if we ever add more clouds behind this command :)
I mean mainly that `cloud connect` wouldn't necessarily be enough for nRF Cloud connection anymore then.

File: samples/nrf9160/modem_shell/prj.conf [hunk 12]
```diff
@@ -128,6 +129,9 @@ CONFIG_MULTICELL_LOCATION_MAX_NEIGHBORS=17
 
 # Cellular positioning services
 CONFIG_MULTICELL_LOCATION_SERVICE_NRF_CLOUD=y
+# Uncomment the following if MQTT is selected, for example when CONFIG_MOSH_CLOUD is set. Otherwise,
+# REST API is used for communication between the device and nRF Cloud.
+#CONFIG_NRF_CLOUD_CELL_POS=y
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/prj.conf:134]: Does MOSH compile if you have `CONFIG_NRF_CLOUD_MQTT` but not `CONFIG_NRF_CLOUD_CELL_POS`. Doesn't need to work but need to see if we should prevent that.

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+static struct k_work notify_pgps_work;
+#endif
+
+static const char cloud_shell_cmd_usage_str[] =
+	"MQTT connection to nRFCloud\n"
+	"Usage: cloud <subcommand>\n"
+	"\n"
+	"Subcommands:\n"
+	"  connect:      Establish MQTT connection to nRFCloud.\n"
+	"  disconnect:   Disconnect from nRFCloud.\n";
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:34]:
```suggestion
static const char cloud_shell_cmd_usage_str[] =
	"MQTT connection to nRF Cloud\n"
	"Usage: cloud <subcommand>\n"
	"\n"
	"Subcommands:\n"
	"  connect:      Establish MQTT connection to nRF Cloud.\n"
	"  disconnect:   Disconnect from nRF Cloud.\n";
```

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+	int ret = 0;
+
+	/* command = argv[0] = "cloud" */
+	/* sub-command = argv[1]       */
+	if (argc < 2) {
+		goto show_usage;
+	} else if (strcmp(argv[1], "connect") == 0) {
+		cloud_connect();
+	} else if (strcmp(argv[1], "disconnect") == 0) {
+		cloud_disconnect();
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:178]: One more thing. We should have a discussion on how to handle subcommands. Whether this manual stuff is fine or should we use shell macros? I think we should move towards shell macros to get auto-fill feature into use.

File: samples/nrf9160/modem_shell/prj.conf (hunk 12 above)
Comment (tokangas) [inline samples/nrf9160/modem_shell/prj.conf:134]: Would it make sense to select this automatically in Kconfig if MOSH_CLOUD and MULTICELL_LOCATION_SERVICE_NRF_CLOUD are enabled?

File: samples/nrf9160/modem_shell/Kconfig (hunk 9 above)
Comment (trantanen) [inline samples/nrf9160/modem_shell/Kconfig:89]: Remove this line as it's default. I've been instructed to remove these few times although I like that it documents that it's intentionally OFF by default.

Comment (trantanen) [reply samples/nrf9160/modem_shell/Kconfig:89]: Already commented by @tokangas

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c (hunk 14 above)
Comment (tokangas) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:178]: Shell subcommands would work nicely in this case, so I think it would make sense to use those instead.

File: samples/nrf9160/modem_shell/prj.conf (hunk 12 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/prj.conf:134]: The answer is no. I get this:
`/home/tor1/work/ncs/nrf/lib/multicell_location/services/nrf_cloud_integration.c:96: undefined reference to `nrf_cloud_cell_pos_request'`

I'll make BUILD_ASSERT to multicell_location library so the error becomes clear. I guess we don't have a way to handle this in multicell Kconfig.

Comment (hiltunent) [reply samples/nrf9160/modem_shell/prj.conf]: Agreed

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c (hunk 14 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c]: Done

File: samples/nrf9160/modem_shell/Kconfig [hunk 15]
```diff
@@ -84,6 +84,13 @@ config MOSH_REST
 	help
 	  REST client for sending and receiving simple REST requests/responses.
 
+config MOSH_CLOUD
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/Kconfig:87]: Documentation update is something that could be already done

File: samples/nrf9160/modem_shell/prj.conf (hunk 12 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/prj.conf]: Dependencies are handled with:
https://github.com/nrfconnect/sdk-nrf/pull/6223
https://github.com/nrfconnect/sdk-nrf/pull/6244
Resolving this comment.

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.h [hunk 16]
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef CLOUD_SHELL_H
+#define CLOUD_SHELL_H
+
+int cloud_shell(const struct shell *shell, size_t argc, char **argv);
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.h:10]: Is this file needed somewhere? At least this function doesn't exist anymore.
'gnss' module seems to have similar issue.

File: samples/nrf9160/modem_shell/prj.conf (hunk 1 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/prj.conf]: Maybe @jhirsi could close this as dependencies are handled with:
https://github.com/nrfconnect/sdk-nrf/pull/6223
https://github.com/nrfconnect/sdk-nrf/pull/6244

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c (hunk 10 above)
Comment (tokangas) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:112]: Maybe not, but the warning trace kind of suggests that it happens automatically. :)

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+{
+	int err;
+
+	struct nrf_cloud_init_param config = {
+		.event_handler = nrf_cloud_event_handler,
+	};
+
+	err = nrf_cloud_init(&config);
+	if (err == -EACCES) {
+		mosh_print("nRF Cloud module already initialized");
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:133]: This print could be removed. It's not necessary information for the user and it will be printed for every "cloud connect" command except the first one, right?

Comment (trantanen) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:133]: But the user knows if he has established the connection already? Or is it possible that an old cloud connection is alive and will be reused?
I don't know some of these details to decide if I agree or disagree :)

Comment (tokangas) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:133]: I was thinking something like "cloud connect", "cloud disconnect" and "cloud connect". :)

There's a separate print when `nrf_cloud_connect()` returns `NRF_CLOUD_CONNECT_RES_ERR_ALREADY_CONNECTED`. This one is different, the user doesn't have to care whether the API has been initialized or not, that should happen under the hood.

Comment (hiltunent) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:133]: I'll remove the print.

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.h (hunk 16 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.h:10]: Good catch, I'll remove cloud_shell.h. I can do similar change for GNSS as well in this PR but in a separate commit. I hope it's ok.

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c (hunk 10 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:112]: You're right, I may have used copy-paste a bit too liberally here :) I'll change the warning message for now. If there is a case for having auto-reconnect, that can be implemented too.

File: samples/nrf9160/modem_shell/prj.conf (hunk 1 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/prj.conf]: Excellent, resolving.

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+static void cmd_cloud_connect(const struct shell *shell, size_t argc, char **argv)
+{
+	int err;
+
+	struct nrf_cloud_init_param config = {
+		.event_handler = nrf_cloud_event_handler,
+	};
+
+	err = nrf_cloud_init(&config);
+	if (err && (err != -EACCES)) {
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:132]: Doesn't this change the behavior with `-EACCES`? Earlier that resulted in the `mosh_print()` call, but now it would get handle by the `else` branch if P-GPS is enabled, which is not intended behavior. Either the `else` branch has to check that `err` is not `-EACCES`, or it may be easier to just add a static bool flag to store whether initialization has been done or not.

File: samples/nrf9160/modem_shell/src/shell.c [hunk 19]
```diff
@@ -46,6 +46,7 @@
 #if defined(CONFIG_MOSH_REST)
 #include "rest_shell.h"
 #endif
+
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/shell.c:49]: **nitpick** You could remove this added empty line.

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c (hunk 18 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c]: flag added

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.h (hunk 16 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.h]: removed

File: samples/nrf9160/modem_shell/Kconfig (hunk 15 above)
[the remark this answers is not in the document]
Comment (hiltunent) [reply samples/nrf9160/modem_shell/Kconfig:87]: Changelog and mosh readme edited. Please take a look. I also added @peknis

File: samples/nrf9160/modem_shell/README.rst [hunk 20]
```diff
@@ -499,6 +499,29 @@ Examples
...
+nRF Cloud
+=========
+
+MoSh command: ``cloud``
+
+nRF Cloud is a platform for providing, among other things, various location services. Modem shell
+enables the user to establish an MQTT connection to nRF Cloud by using :ref:`lib_nrf_cloud`.
+Currently, ``cloud`` command is useful mostly when utilizing the Location tool and MQTT is the
+desired transport protocol. However, there is no limit on what nRF Cloud services can be used once
+the MQTT connection is established.
```
Comment (peknis) [inline samples/nrf9160/modem_shell/README.rst:513]:
```suggestion
nRF Cloud is a platform for providing, among other things, various location services.
Modem shell enables you to establish an MQTT connection to nRF Cloud using the :ref:`lib_nrf_cloud` library.
Currently, the ``cloud`` command is useful mostly when using the location services and MQTT is the desired transport protocol.
However, you can use any nRF Cloud services once the MQTT connection is established.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: samples/nrf9160/modem_shell/README.rst (hunk 20 above)
Comment (hiltunent) [reply samples/nrf9160/modem_shell/README.rst]: Thanks. I updated also the example...it was a bit too trivial.

File: samples/nrf9160/modem_shell/README.rst [hunk 21]
```diff
@@ -499,6 +499,29 @@ Examples
...
 
+----
+
+nRF Cloud
+=========
+
+MoSh command: ``cloud``
+
+nRF Cloud is a platform for providing, among other things, various location services.
+Modem shell enables you to establish an MQTT connection to nRF Cloud using the :ref:`lib_nrf_cloud` library.
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/README.rst:510]:
```suggestion
Modem Shell enables you to establish an MQTT connection to nRF Cloud using the :ref:`lib_nrf_cloud` library.
```

**Review by tokangas — APPROVED** [review]

File: samples/nrf9160/modem_shell/README.rst [hunk 22]
```diff
@@ -499,6 +499,29 @@ Examples
...
+Currently, the ``cloud`` command is useful mostly when using the location services and MQTT is the desired transport protocol.
+However, you can use any nRF Cloud services once the MQTT connection is established.
+
+Examples
+--------
+
+* Establish the connection to nRF Cloud, request cell based location of the device, and disconnect when ready:
```
Comment (peknis) [inline samples/nrf9160/modem_shell/README.rst:517]:
```suggestion
* Establish the connection to nRF Cloud, request the cell-based location of the device, and disconnect when ready:
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: samples/nrf9160/modem_shell/prj.conf [hunk 23]
```diff
@@ -13,6 +13,7 @@ CONFIG_MOSH_LINK=y
 CONFIG_MOSH_SMS=y
 CONFIG_MOSH_GNSS=y
 CONFIG_MOSH_LOCATION=y
+#CONFIG_MOSH_CLOUD=y
```
[the remark this answers is not in the document]
Comment (hiltunent) [reply samples/nrf9160/modem_shell/prj.conf:16]: Thanks. I added the basic MQTT overlay which is now in sample.yaml. The MQTT+REST is there too but I did not create overlay for it.

**Review by peknis — APPROVED** [review]

File: samples/nrf9160/modem_shell/README.rst [hunk 24]
```diff
@@ -499,6 +499,29 @@ Examples
 
      rest -d example.com -l 1024 -m head -H "X-foo1: bar1\x0D\x0A" -H "X-foo2: bar2\x0D\x0A"
 
+----
+
+nRF Cloud
```
[the remark this answers is not in the document]
Comment (hiltunent) [reply samples/nrf9160/modem_shell/README.rst]: Changed to simply "Cloud"

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+	shell_help(shell);
+
+	return ret;
+}
+
+static void cloud_reconnect_work_fn(struct k_work *work)
+{
+	int err = nrf_cloud_connect(NULL);
+
+	if (err == NRF_CLOUD_CONNECT_RES_ERR_ALREADY_CONNECTED) {
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:47]: Quickly what we discussed. This could be organized so that first comparison is for NRF_CLOUD_CONNECT_RES_SUCCESS so it would work even if enum values would change (not that it would happen but I think the code is more self-documentary that way).
```
if (err == NRF_CLOUD_CONNECT_RES_SUCCESS) {
	mosh_print("Connection to nRF Cloud re-established");
} else if (err == NRF_CLOUD_CONNECT_RES_ERR_ALREADY_CONNECTED) {
	mosh_print("nRF Cloud connection already established");
} else {
	mosh_error("nrf_cloud_connect, error: %d", err);
}
```

And the same code is in `cmd_cloud_connect()` with very little text changes. If the disconnect event seems to be coming, I think we could consider `cmd_cloud_connect()` to call this function.

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+		}
+
+		initialized = true;
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+		k_work_init(&notify_pgps_work, notify_pgps);
+#endif
+		k_work_init_delayable(&cloud_reconnect_work, cloud_reconnect_work_fn);
+	}
+
+	k_work_reschedule(&cloud_reconnect_work, K_NO_WAIT);
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:165]: This was changed to a scheduled work. I don't know if calling the work function would work in practice but I'll leave it to you which way to go:
`cloud_reconnect_work_fn(NULL);`

Comment (trantanen) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:165]: I think I've called work functions directly without issues as long as it doesn't need information from the k_work parameter. Anyway, it was more like a note that you could still call it directly but up to you which way you want to do it.

**Review by trantanen — APPROVED** [review]

**Review by jhirsi — APPROVED** [review]

File: samples/nrf9160/modem_shell/src/cloud/cloud_shell.c (hunk 26 above)
Comment (tokangas) [reply samples/nrf9160/modem_shell/src/cloud/cloud_shell.c:165]: I haven't called work functions directly either, although technically possible. When called directly, the function is executed in a different thread. There are a few consequences, like the thread stack size is likely different and in some cases it might be possible that the function would be executed in two different threads at the same time. Also, when searching the code for places where the work is used, a direct work function call may easily be missed. I think it would be better to be consistent and always execute works using the `k_work` functions.

**Review by rlubos — APPROVED** [review]

**Comment by plskeggs** [comment]
@hiltunent I noticed an issue with your change to recover from a disconnection -- it does not allow the `cloud disconnect` command to remain disconnected.  It forces an automatic reconnection despite the user's request to disconnect.

**Comment by rlubos** [comment]
@hiltunent The CI issue looks like a valid one, please have a look.

**Comment by hiltunent** [comment]
> @hiltunent The CI issue looks like a valid one, please have a look.

Yes, I fixed this now. The CI will likely still not pass until https://github.com/nrfconnect/sdk-nrf/pull/6244 is merged.

**Comment by hiltunent** [comment]
> @hiltunent I noticed an issue with your change to recover from a disconnection -- it does not allow the `cloud disconnect` command to remain disconnected. It forces an automatic reconnection despite the user's request to disconnect.

This is rather embarrassing. I just fixed it.

File: samples/nrf9160/modem_shell/README.rst [hunk 27]
```diff
@@ -499,6 +499,29 @@ Examples
 
      rest -d example.com -l 1024 -m head -H "X-foo1: bar1\x0D\x0A" -H "X-foo2: bar2\x0D\x0A"
 
+----
+
+Cloud
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/README.rst:504]: We are missing compilation instructions from `Building and running` section so there should be a new subsection like for other overlays.

Comment (hiltunent) [reply samples/nrf9160/modem_shell/README.rst:504]: Done!

**Review by plskeggs — APPROVED** [review]
Looks good.  Please confirm it builds when enabling this with `overlay-cloud_mqtt.conf` -- I was having some trouble but could be due to stacking my PR on top (and I'm done for the year...)

**Comment by hiltunent** [comment]
> Looks good. Please confirm it builds when enabling this with `overlay-cloud_mqtt.conf` -- I was having some trouble but could be due to stacking my PR #6309 on top (and I'm done for the year...)

Yes sir, you most likely you need to have  #6244 as well. At least the CI is passing no so I am confident it is ok.
